model/src/main/java/org/projectnessie/model/Branch.java

```diff
@@ -31,7 +32,7 @@
 @JsonSerialize(as = ImmutableBranch.class)
 @JsonDeserialize(as = ImmutableBranch.class)
 @JsonTypeName("BRANCH")
-public interface Branch extends Reference {
+public interface Branch extends Reference, Serializable {
```

Comment:
Keeping classes of `model` as `Serializable` is not good. Hence, I used JSON string instead of actual class for the references.

I think same problem comes for `Tag` also, currently we don't have test case that creates a tag and drops it and use it for GC. Hence, it looks like problem is not there for `Tag` 

Also keeping `Serializable` with `Immutables` is a problem with spark code generation (which happens with dataframe), code generation will fail because these Immutables classes don't have public consutructor, cannot add public constructor as we are using `@Value.Check` annotation.

Reply:
yes, using immutables is problematic. One solution would be to create separate data classes that are used only in spark and create a mapper logic between Spark classes and immutables.

Reply:
How efficient it would be compared current JSON serialization? I agree that Kryo is faster but this code is looking complex and adding mapper will bring extra complexity. 

So, I am happy with the current JSON serialization considering the complexity. Let's wait for the others opinion.

Reply:
The kryo should be orders of magnitude faster compared to json (see https://blog.softwaremill.com/the-best-serialization-strategy-for-event-sourcing-9321c299632b) and also it requires a lot less space for storing each object in memory, so I think we should definitely use kryo instead of JSON.

Reply:
In Spark, it is a good pattern to model the domain using case classes (when writing in scala) and carry only the parts of data that are indeed needed for processing. Since we are using Java we cannot use case classes but we can model our problem and keep it contained to the spark module. This is the same pattern as for rest APIs and DTOs. The entity (persistence) model should not leak into a REST api that is why the DTOs are used. For big data processing, we should also have a dedicated model classes to not tight couple spark processing to our model.